### PR TITLE
Project shower vertices onto PCA primary axis.

### PR DESCRIPTION
**Pull request by StevenGreen1**
Alter the shower start position by projecting the vertex onto the PCA axis. 

The following images demonstrate this in action.  The red dot is the initial vertex, blue is the centroid and green is the new vertex.  The green line is the PCA primary axis.

Once we've reviewed this change on GitHub, I'll cherry-pick these changes and add them to a redmine feature branch. 

Screenshot 2019-09-06 at 15 53 14
Screenshot 2019-09-06 at 15 52 35
Screenshot 2019-09-06 at 15 51 52
Screenshot 2019-09-06 at 15 51 01

### CODE REVIEW

File: larpandora/LArPandoraEventBuilding/LArPandoraShowerCreation_module.cc [hunk 1]
```diff
@@ -182,9 +182,17 @@ void LArPandoraShowerCreation::produce(art::Event &evt)
...
         {
-            // Ensure successful creation of all structures before placing results in output containers
-            const lar_content::LArShowerPCA larShowerPCA(lar_content::LArPfoHelper::GetPrincipalComponents(cartesianPointVector, vertexPosition));
-            const recob::Shower shower(LArPandoraShowerCreation::BuildShower(showerCounter++, larShowerPCA, vertexPosition));
+            // Access centroid of shower via this method
+            const lar_content::LArShowerPCA initialLArShowerPCA(lar_content::LArPfoHelper::GetPrincipalComponents(cartesianPointVector, vertexPosition));
+
+            // Project the PFParticle vertex onto the PCA axis
+            const pandora::CartesianVector centroid(initialLArShowerPCA.GetCentroid());
+            const pandora::CartesianVector primaryAxis(initialLArShowerPCA.GetPrimaryAxis());
```
[the remark this answers is not in the document]
Comment (StevenGreen1) [reply larpandora/LArPandoraEventBuilding/LArPandoraShowerCreation_module.cc]: Good idea, will apply.

File: larpandora/LArPandoraEventBuilding/LArPandoraShowerCreation_module.cc [hunk 2]
```diff
@@ -182,9 +182,17 @@ void LArPandoraShowerCreation::produce(art::Event &evt)
...
+            const pandora::CartesianVector projectedVertexPosition(centroid - primaryAxis.GetUnitVector() * (centroid - vertexPosition).GetDotProduct(primaryAxis));
+
+            // Ensure successful creation of all structures before placing results in output containers, remaking LArShowerPCA with updated vertex
+            const lar_content::LArShowerPCA larShowerPCA(lar_content::LArPfoHelper::GetPrincipalComponents(cartesianPointVector, projectedVertexPosition));
```
[the remark this answers is not in the document]
Comment (StevenGreen1) [reply larpandora/LArPandoraEventBuilding/LArPandoraShowerCreation_module.cc]: Yeah you're right regarding the eigenvectors and eigenvalues.  In principle for most cases, I don't think this should change either as the vertex and the projected vertex should be on the same side of the primary axis (w.r.t. the centroid).  The only reason I reran it here was to be extra safe.  I think a new constructor sounds good though (although adds a LArPandoraContent change too).  I'll add that in now too and use it.

**Comment by StevenGreen1** [comment]
Hi John, 

Thanks for the comments.  For the top picture, a few hits are missing to the right of what's shown.  I zoomed in to make the vertices clear, but from what I've seen the PCA axes look good to me.  I'll apply the changes discussed above and update the PR shortly. 

Cheers
Steve

File: larpandora/LArPandoraEventBuilding/LArPandoraShowerCreation_module.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (PandoraPFA) [reply larpandora/LArPandoraEventBuilding/LArPandoraShowerCreation_module.cc]: The constructor is already there, so no need for a LArContent change, which helps!

**Comment by PandoraPFA** [comment]
Great, good to know the axes look visually OK - thanks!

File: larpandora/LArPandoraEventBuilding/LArPandoraShowerCreation_module.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (StevenGreen1) [reply larpandora/LArPandoraEventBuilding/LArPandoraShowerCreation_module.cc]: Ah I thought you meant a constructor that took in a LArShowerPCA and a vertex that would do the sign flipping part only.  I've used the constructor taking in all the individual axes though and it looks reasonably clean to me now I think though.

**Comment by StevenGreen1** [comment]
Great, thanks for the comments.  I'll get a feature branch ready and email Lyn.  The cosmetic comment is sensible for sure, but if you're happy for now we can leave things as they are.  Cheers, Steve

**Comment by StevenGreen1** [comment]
Merged into LArSoft.

**Pull request closed without merging**